File: python/fuzz/targets/receipt_der.py
"""The whole legacy-receipt path on DER bytes: the ``asn1crypto`` CMS walk,
the receipt payload parse, chain building and the RSA signature check.

This is the target that reaches ``asn1crypto``'s ``ContentInfo`` /
``SignedData`` / SET decoding through the entry point a consumer calls, which
is the only way that parser is reachable at all — the port has no public
ASN.1 surface of its own.

Three invariants, the same ones the Go, Rust and Node ports state:

  * nothing crashes the interpreter;
  * a failure is a ``VerificationError``, never a foreign error type;
  * a receipt that verifies was accepted *because of* the anchors, proven by
    re-running it against an unrelated anchor set and requiring failure.
    Without that third one a fuzz target can only find crashes, never
    "accepts what it should not".
"""

from harness import (
    RECEIPT_ANCHORS,
    UNRELATED_ANCHORS,
    InvariantViolation,
    require_verification_error,
    run,
    verify_receipt_core,
)


def one_input(data: bytes) -> None:
    try:
        verify_receipt_core(data, RECEIPT_ANCHORS)
    except Exception as error:
        require_verification_error(error, "verify_receipt_core")
        return
    try:
        verify_receipt_core(data, UNRELATED_ANCHORS)
    except Exception as error:
        require_verification_error(error, "verify_receipt_core against the unrelated anchors")
        return
    raise InvariantViolation(
        "this input verifies against an unrelated anchor set too, "
        "so the anchors are not being enforced"
    )


if __name__ == "__main__":
    run(one_input)

File: dotnet/fuzz/Targets/ReceiptDer.cs
using System;
using System.Collections.Generic;
using System.Security.Cryptography.X509Certificates;
using ApplePurchaseReceiptVerifier.Receipt;

namespace ApplePurchaseReceiptVerifier.Fuzz.Targets
{
    /// <summary>
    /// The whole legacy-receipt path on DER bytes: the structural pre-scan,
    /// then <c>SignedCms</c> (BER), the payload parse, the chain build and the
    /// signature check.
    /// </summary>
    /// <remarks>
    /// The invariants are the Go port's three. Nothing may escape but a
    /// <see cref="VerificationException"/>; the pre-scan and the full path must
    /// agree on the certificate bound; and an accepted receipt is accepted
    /// <em>because of</em> the anchors, proven by re-running it against an
    /// unrelated anchor set and requiring failure. Without that last one a
    /// fuzzer finds crashes and never "accepts what it should not".
    /// <para>The trusted set is the pinned Apple roots plus the generated
    /// fixture receipt root, so the shared fixtures and the two public Apple
    /// receipts get past the chain check and the fuzzer can explore what lies
    /// beyond it. The unrelated set is the fixture <em>JWS</em> root.</para>
    /// </remarks>
    internal sealed class ReceiptDer : IDisposable
    {
        private readonly List<X509Certificate2> _trusted;
        private readonly List<X509Certificate2> _unrelated;

        internal ReceiptDer()
        {
            _trusted = new List<X509Certificate2>(AppleRootCertificates.ReceiptRoots())
            {
                Fixtures.ReceiptRoot(),
            };
            _unrelated = new List<X509Certificate2> { Fixtures.JwsRoot() };
        }

        internal void Run(ReadOnlySpan<byte> data)
        {
            byte[] der = data.ToArray();

            // The pre-scan runs before SignedCms sees anything, so fuzz it
            // both as the library calls it and on its own: its refusal to
            // decode a certificate is what keeps a certificate flood cheap,
            // and a leak here is a leak before any bound applies.
            int embedded;
            try
            {
                embedded = Internals.CmsPreScan(der, Internals.MaxEmbeddedCertificates);
            }
            catch (Exception e)
            {
                Invariant.Contained("CmsPreScan.Scan", e);
                embedded = -1;
            }

            AppReceipt receipt;
            try
            {
                receipt = ReceiptVerifier.VerifyReceiptCore(der, _trusted);
            }
            catch (Exception e)
            {
                Invariant.Contained("VerifyReceiptCore", e);
                return;
            }

            Invariant.Require(receipt is not null, "a receipt that verified came back null");
            Invariant.Require(
                embedded >= 0 && embedded <= Internals.MaxEmbeddedCertificates,
                $"a receipt verified while the pre-scan counted {embedded} embedded certificates");

            try
            {
                ReceiptVerifier.VerifyReceiptCore(der, _unrelated);
            }
            catch (VerificationException)
            {
                return;
            }

            throw new InvariantException(
                "this input verifies against an unrelated anchor set too, "
                + "so the anchors are not being enforced");
        }

        public void Dispose()
        {
            foreach (X509Certificate2 anchor in _trusted)
            {
                anchor.Dispose();
            }

            foreach (X509Certificate2 anchor in _unrelated)
            {
                anchor.Dispose();
            }
        }
    }
}
